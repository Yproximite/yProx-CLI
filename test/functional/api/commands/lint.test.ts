import { readFixture } from '../../../fixtures';
import { createFakeEnv } from '../../fake-env';

const files = {
  'package.json': readFixture('modern-project/package.json'),
  'yarn.lock': readFixture('modern-project/yarn.lock'),
  '.eslintignore': readFixture('modern-project/.eslintignore'),
  '.eslintrc': readFixture('modern-project/.eslintrc'),
  '.stylelintrc': readFixture('modern-project/.stylelintrc'),
  // rollup
  'src/components/button/Button.vue': readFixture('modern-project/src/components/button/Button.vue'),
  'src/components/button/index.js': readFixture('modern-project/src/components/button/index.js'),
  // js
  'src/js/bar.js': readFixture('modern-project/src/js/bar.js'),
  'src/js/foo.js': readFixture('modern-project/src/js/foo.js'),
  // images
  'src/images/guts-white-hair.png': readFixture('modern-project/src/images/guts-white-hair.png', null),
  'src/images/jax.jpg': readFixture('modern-project/src/images/jax.jpg', null),
  'src/images/golfer.gif': readFixture('modern-project/src/images/golfer.gif', null),
  'src/images/uk.svg': readFixture('modern-project/src/images/uk.svg', null),
};

describe('command: lint', () => {
  let oldEnv = process.env;
  beforeEach(() => {
    oldEnv = process.env;
    console.log = jest.fn();
    console.info = jest.fn();
    console.error = jest.fn();
    // @ts-ignore
    process.exit = jest.fn();
  });
  afterEach(() => {
    process.env = oldEnv;
    // @ts-ignore
    console.log.mockRestore();
    // @ts-ignore
    console.info.mockRestore();
    // @ts-ignore
    console.error.mockRestore();
    // @ts-ignore
    process.exit.mockRestore();
  });

  describe('lint (but not fix)', () => {
    it('should lint entries and fails', async () => {
      const { api, run, writeFile, cleanup } = await createFakeEnv({ files });

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint');

      // The first entry handled by yprox-cli is a `rollup` entry.
      // Since we lint files grouped by handlers, it means that we are linting `rollup` files first
      expect(console.error).toHaveBeenCalledWith('[08:30:00] error :: rollup (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).toHaveBeenCalledWith(1);

      // But if we fix the incriminated file...
      await writeFile('src/components/button/index.js', "console.log('ESLint will not fails here.')");
      await api.executeCommand('lint');

      // Then it's another linter which fails
      expect(console.error).toHaveBeenCalledWith('[08:30:00] error :: js (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).toHaveBeenCalledWith(1);

      await cleanup();
    }, 10000);

    it('should lint Rollup entries only and fails', async () => {
      const { api, run, cleanup } = await createFakeEnv({ files });

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint', {
        'filter:handler': 'rollup',
      });

      expect(console.error).toHaveBeenCalledWith('[08:30:00] error :: rollup (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).toHaveBeenCalledWith(1);

      await cleanup();
    }, 10000);

    it('should lint JS entries only and fails', async () => {
      const { api, run, cleanup } = await createFakeEnv({ files });

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint', {
        'filter:handler': 'js',
      });

      expect(console.error).toHaveBeenCalledWith('[08:30:00] error :: js (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).toHaveBeenCalledWith(1);

      await cleanup();
    }, 10000);
  });

  describe('lint (and fix)', () => {
    it('should lint and fix entries', async () => {
      const { api, run, cleanup } = await createFakeEnv({ files });

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint', {
        fix: true,
      });

      expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: Your JavaScript is clean ✨');
      expect(console.error).not.toHaveBeenCalledWith('[08:30:00] error :: rollup (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).not.toHaveBeenCalledWith(1);

      await cleanup();
    }, 10000);

    it('should lint and fix Rollup entries only', async () => {
      const { api, run, cleanup, readFile } = await createFakeEnv({ files });

      const fileContent = await readFile('src/components/button/index.js');
      expect(fileContent).toMatchSnapshot('button/index.js before lint');

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint', {
        'filter:handler': 'rollup',
        fix: true,
      });

      expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: Your JavaScript is clean ✨');
      expect(console.error).not.toHaveBeenCalledWith('[08:30:00] error :: rollup (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).not.toHaveBeenCalledWith(1);

      const newFileContent = await readFile('src/components/button/index.js');
      expect(newFileContent).toMatchSnapshot('button/index.js after lint');
      expect(fileContent).not.toBe(newFileContent);

      await cleanup();
    }, 10000);

    it('should lint and fix JS entries only', async () => {
      const { api, run, cleanup, readFile } = await createFakeEnv({ files });

      const fileContent = await readFile('src/js/bar.js');
      expect(fileContent).toMatchSnapshot('js/bar.js before lint');

      await run('yarn install --frozen-lockfile');
      await api.executeCommand('lint', {
        'filter:handler': 'js',
        fix: true,
      });

      expect(console.info).toHaveBeenCalledWith('[08:30:00] info :: Your JavaScript is clean ✨');
      expect(console.error).not.toHaveBeenCalledWith('[08:30:00] error :: js (lint) :: Your JavaScript is not clean, stopping.');
      expect(process.exit).not.toHaveBeenCalledWith(1);

      const newFileContent = await readFile('src/js/bar.js');
      expect(newFileContent).toMatchSnapshot('js/bar.js after lint');
      expect(fileContent).not.toBe(newFileContent);

      await cleanup();
    }, 10000);
  });
});
