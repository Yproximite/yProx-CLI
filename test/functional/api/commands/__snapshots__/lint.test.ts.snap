// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command: lint CSS & Sass should fix linting issues: bootstrap-grid.scss before linting 1`] = `
"@import \\"~bootstrap/scss/bootstrap-grid\\";;
"
`;

exports[`command: lint CSS & Sass should fix linting issues: src/button/Button.vue after linting 1`] = `
"@import \\"~bootstrap/scss/bootstrap-grid\\";
"
`;

exports[`command: lint CSS & Sass should fix linting issues: src/button/index.js after linting 1`] = `
".foo {
  color: blue;
}
"
`;

exports[`command: lint CSS & Sass should fix linting issues: style.css before linting 1`] = `
".foo {
  color: blue;;
}
"
`;

exports[`command: lint JavaScript should lint and fix linting issues: src/es6.js after linting 1`] = `
"const constant = 'abc';
console.log(\`The constant value: \${constant}\`);

let arr = [1, 2, 3];
arr = [...arr];

let obj = { a: '1', b: '2', c: '3' };
obj = { ...obj };

const myFunction = (arg1, arg2, ...otherArgs) => {
  console.log({
    arg1,
    arg2,
    otherArgs,
  });
};
"
`;

exports[`command: lint JavaScript should lint and fix linting issues: src/es6.js before linting 1`] = `
"const constant = 'abc';
console.log(\`The constant value: \${constant}\`);;

let arr = [1, 2, 3];
arr = [...arr];

let obj = { a: '1', b: '2', c: '3' };
obj = { ...obj };

const myFunction = (arg1, arg2, ...otherArgs) => {
  console.log({
    arg1,
    arg2,
    otherArgs,
  });
};
"
`;

exports[`command: lint JavaScript should lint and fix linting issues: src/hello-world.js after linting 1`] = `
"console.log('Hello world from!');
"
`;

exports[`command: lint JavaScript should lint and fix linting issues: src/hello-world.js before linting 1`] = `
"console.log('Hello world from!');;
"
`;

exports[`command: lint Vue should fix linting issues: src/button/Button.vue after linting 1`] = `
"<template>
  <button
    @click=\\"$emit('click')\\"
  >
    {{ text }}
  </button>
</template>

<script>
export default {
  name: \\"Button\\",
  props: {
    text: {
      type: String,
      required: true,
    }
  },
  data() {
    return {
      'i-should': 'be after \\"props\\" definition'
    }
  },

  created() {
    console.log('Hello from Button.vue!')
    console.log({ ...this.$props })

    if (this.text.includes('foobar')) {
      console.log('\\"text\\" contains \\"foobar\\"');
    }
  }
};
</script>
"
`;

exports[`command: lint Vue should fix linting issues: src/button/Button.vue before linting 1`] = `
"<template>
  <button @click=\\"$emit('click')\\"
  >
    {{ text }}
  </button>
</template>

<script>
export default {
  name: \\"Button\\",
  data() {
    return {
      'i-should': 'be after \\"props\\" definition'
    }
  },
  props: {
    text: {
      type: String,
      required: true,
    }
  },

  created() {
    console.log('Hello from Button.vue!')
    console.log({ ...this.$props })

    if (this.text.includes('foobar')) {
      console.log('\\"text\\" contains \\"foobar\\"');
    }
  }
};
</script>
"
`;

exports[`command: lint Vue should fix linting issues: src/button/index.js after linting 1`] = `
"import Vue from 'vue';
import Button from './Button';

Vue.component('y-button', Button);

console.log(\\"Hello from index.js!\\");
"
`;

exports[`command: lint Vue should fix linting issues: src/button/index.js before linting 1`] = `
"import Vue from 'vue';
import Button from './Button';

Vue.component('y-button', Button);

console.log(\\"Hello from index.js!\\");;
"
`;
